### DISCUSSION TITLE
How to parse OL events to Purview / Apache Atlas entities?

### DISCUSSION THREAD

**Discussion by plaksnor** [OpenLineage/OpenLineage#1990] [Ideas]
Hi,
I'm very excited about this project! Great work guys! But I really could get some help over here to make things work!

At my organization we work with Synapse Notebooks and Microsoft Purview. Synapse Notebooks are not recognized by Purview, so we consider using [PyApacheAtlas](https://github.com/wjohnson/pyapacheatlas) to load missing assets and its metadata. Besides that, this project could help reading missing metadata as well. However, we're not allowed to use Azure Functions, Kafka, Event Grid or other message brokers.
One other thing to mention: Synapse Notebook users/processes are not allowed to use Service Principals to make a direct connection with Purview: they simply could destroy everything, which I understand. So my hands are very tied.

### The idea:
1. OpenLineage-spark.jar (transport-type: console) writes events in the Apache Spark log (stderr logs).
2. In the same session, the user or process runs additional code to send its stderr logs to a central location **
3. The logs get parsed to Apache Atlas objects/entities

In another process with elevated privileges:
4. Objects will get processed with PyApacheAtlas or other code to use the Purview API to get data in Purview.

** Step 2 and 3 can be switched though, but it may seems to be logic that we need settings from the current session to get additional info.
** I'm trying to parse the OpenLineage events by processing the stderr logs directly from the same Synapse Notebook session as which creates the events. This approach has a benefit for other notebook users in the future, because this way they will get more flexibility by being able to choose themselves when they want to publish their entities and processes to Purview. We actually don't want all notebook logs to be parsed to Purview. Now users just simply need to embed 1 additional line of code by using a magic command (%run) which includes this notebook for publishing assets in Purview.

So in short:
OpenLineage-spark.jar -> stderr logs -> central location to pick data up by another proces -> PyApacheAtlas -> Purview

I'm running Apache Spark 3.3 with the [openlineage-spark-0.29.2.jar](https://repo1.maven.org/maven2/io/openlineage/openlineage-spark/0.29.2/openlineage-spark-0.29.2.jar) package and use 'console' as the transport type. This way I get CustomTransport events in the console (Apache Spark log: stderr). Working with PyApacheAtlas is also no problem and we do have some scripts to create custom entities or create AtlasEntity or AtlasProcess calls. And the Excel functionality is also very handy.

### TLDR;
How do I properly parse OL events with Python to entities ready to be send to Purview / Apache Atlas?
Synapse Notebook users/processes are not in the position to create a live connection to the Purview API, so my end goal is parsing the data to Purview (or Apache Atlas) objects via a database, which can be used later on in a proces with elevated priviliges which will send all the data.

I prefer using an existing library that has already covered most of the parsing functionality, but it's no problem tying things together as long as things are easy to maintain. I'm open for other suggestions and I'm looking forward to any new ideas!

**Comment by mobuchowski** [comment]
@plaksnor custom transports are separate thing from console transport. And, as I understood your limitations, you can't use custom transport regardless. Besides that, what you wrote makes sense. 

I assume that your problem is just mapping data from OpenLineage events to Atlas/Purview entities? I'm not sure if there is anything generic now - that given OpenLineage event will give you Atlas entities - but I'm very far from knowing Atlas or it's ecosystem well. However, OpenLineage events are just json - in the simplest version, you can just parse them with `json.loads` and iterate over datasets, transforming them as you like.

**Reply by plaksnor** [reply]
@mobuchowski thanks for your response.
Yeah my problem is 'just' mapping data from OpenLineage events to Atlas/Purview entities. I mean, parsing json is not that hard if you know what you're looking for, but maybe someone has already made a custom transport type and parsing functionality to Atlas/Purview or something like that. Could be such a waste of time if I'm going to rebuild this.

Already found an [Azure function](https://github.com/microsoft/DataLineage/blob/main/sparklin/BlobTriggerFuncApp/BlobTriggerFunction/Synapse_JsonParser.py) in the [SparkLin repo](https://github.com/microsoft/DataLineage/tree/main/sparklin) which contains a lot of logic I could reuse, but it still leads me to a lot of missing metadata. I'll try to make it work, but if someone else has better ideas, please keep me posted.

**Reply by mobuchowski** [reply]
>[SparkLin repo](https://github.com/microsoft/DataLineage/tree/main/sparklin)

Interesting - that OL version in there is super super old though.

> I mean, parsing json is not that hard if you know what you're looking for, but maybe someone has already made a custom transport type and parsing functionality to Atlas/Purview or something like that.

If you wanted to write something else to console than OL events - for example transforming those events in Java rather than Python - I think looking at custom transport would be a good idea. On the other hand, this seems to be task better suited to Python, so custom transport can't help you.

**Reply by plaksnor** [reply]
Yes their OL version is very old and so is their code I guess. Seems like I'm facing compatibility issues here. So at the moment I'm trying to stick with the latest OL version and adjust the SparkLin code in order to recognize the OL events. This way I could try to create the expected outcome.
I could also use the older OL version (0.4.), but then I should probably use older Spark versions. Not the prefered way if you'd ask me.

For now I'll use the OL console transport type, since this type of output suits te best for json parsing in Python from within Azure Synapse Notebooks directly.

**Reply by mobuchowski** [reply]
@plaksnor you can't even use this old version, as there was no transport feature yet, only http 😉

**Reply by plaksnor** [reply]
ok tnx, good to know :)
